---
title: ggGWAS reasoning
author: Sina Rueger
date: "`r Sys.Date()`"
show_toc: true
slug: gggwas-reasoning
output: 
  rmarkdown::html_vignette:
      toc: yes
vignette: >
  %\VignetteIndexEntry{ggGWAS Reasoning}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---


```{r setup, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

Why bother with ggplot2 extension for GWAS summary statistic data visualisations? 

This vignette lays out the [motivation](#reasons) why having a ggplot2 extension is needed. But it also discusses existing [alternatives](#alternatives) and ways to create a ggplot2-like [Q-Q plot](#ways-of-making-a-ggplot2-like-qqplot-plot) and [Manhattanplot](#ways-of-making-a-ggplot2-like-manhattan-plot). 


## GWAS data visualisations

There are two main data visualisations^[There are also other tools used, such as locuszoom plots, but these are harder to implement because they need access to external information, and we will skip these for now.] that are done after the GWAS is run^[We will not discuss their usefulness here.]: 

- [Q-Q plot](https://en.wikipedia.org/wiki/Q%E2%80%93Q_plot): checks the deviation of the P-value distribution from the uniform distribution. 

- [Manhattan plot](https://en.wikipedia.org/wiki/Manhattan_plot): shows the P-value or summary statistic distribution along the chomosomal position. 

<img src="https://user-images.githubusercontent.com/4454726/47022176-323fb600-d15d-11e8-892b-152816ce9574.png" alt="QQplot" height="249">

<p><a href="https://commons.wikimedia.org/wiki/File:Manhattan_Plot.png#/media/File:Manhattan_Plot.png"><img src="https://upload.wikimedia.org/wikipedia/commons/1/12/Manhattan_Plot.png" alt="Manhattan Plot.png" width="640" height="249"></a><br>By M. Kamran Ikram et al - Ikram MK et al (2010) Four Novel Loci (19q13, 6q24, 12q24, and 5q14) Influence the Microcirculation In Vivo. PLoS Genet. 2010 Oct 28;6(10):e1001184. <a href="https://en.wikipedia.org/wiki/Digital_object_identifier" class="extiw" title="w:Digital object identifier">doi</a>:<a rel="nofollow" class="external text" href="https://doi.org/10.1371%2Fjournal.pgen.1001184.g001">10.1371/journal.pgen.1001184.g001</a>, <a href="https://creativecommons.org/licenses/by/2.5" title="Creative Commons Attribution 2.5">CC BY 2.5</a>, <a href="https://commons.wikimedia.org/w/index.php?curid=18056138">Link</a></p>


## Alternatives

Both plotting types can be applied in R with various functions. Just to name a few, here are some links which have also served as inspiration:

- `qqman`: R package for GWAS results: http://www.gettinggeneticsdone.com/2014/05/qqman-r-package-for-qq-and-manhattan-plots-for-gwas-results.html
- `ggman`: R package with ggplot2 wrapper with some neat hacks https://github.com/mkanai/ggman
- Recipe to create Manhattanplot in R with `qqman`, `ggplot2` and `plotly`: https://www.r-graph-gallery.com/wp-content/uploads/2018/02/Manhattan_plot_in_R.html
- Recipe for QQplots in `lattice` (for historical reasons) https://genome.sph.umich.edu/wiki/Code_Sample:_Generating_QQ_Plots_in_R
- Fast QQplot implementation in `ramwas` package: https://rdrr.io/bioc/ramwas/man/qqplotFast.html
- DYI `ggplot2` Manhattan plots: https://github.com/pcgoddard/Burchardlab_Tutorials/wiki/GGplot2-Manhattan-Plot-Function
- `ggbio` package, a toolkit for genomic data (`plotGrandLinear` is an implementation for a Manhattan plot, uses `GRanges` objects, ggplot2 wrapper): https://bioconductor.org/packages/release/bioc/html/ggbio.html 
- `ggman`, wrapper (but no ggplot2 extension) for Manhattan plots: https://github.com/drveera/ggman
- https://www.gettinggeneticsdone.com/2010/07/qq-plots-of-p-values-in-r-using-base.html


An often used package is [qqman](https://rdrr.io/cran/qqman/man/qq.html), that has a function for Q-Q plots `qqman::qq` and Manhattan plots `qqman::manhattan`. 

And there are various ggplot2 wrappers (`ggbio`, `ggman`), but not actual extensions.



## Reasons

The problem is two fold: 

1. GWAS summary statistics usually contains around 1 mio of rows, so plotting becomes a lengthy process. 
2. All existing packages use the base plotting functions, making it difficult to apply any functionalities that `ggplot2` can (e.g. facetting, colouring).

The idea is therefore to create a ggplot2 extension that would facilitate all this: 1) fast plotting (through hexagons and filtering) and 2) building `ggplot2` geoms and stats. 

By writing a ggplot2 extension, we can inherit lots of the default ggplot2 functionalities and shorten the input. 

## Ways of making a ggplot2-like Q-Q plot plot

Let's say we have GWAS summary statistics for a number of SNPs. We call this data `gwas.summarystats`: for a number of SNPs listed by row, we know the SNP identifier (`SNP`) and the P-value (`P`). 

```
SNP     P
rs3342  1e-2
rs83    1e-2
...     ...
```

For illustration purposes, we will use the summary statistics dataset `GWAS.utils::giant`.

```{r}
library(GWAS.utils)
skimr::skim(giant)


```

### 1. qqman::qq

`qqman::qq` is straightforward to use: just pass on the P-value column.

```{r, fig.height = 5, fig.width = 5}
qqman::qq(giant$P)
```

### 2. ggplot2::geom_qq

There is a [`geom_qq`](https://ggplot2.tidyverse.org/reference/geom_qq.html) in ggplot2 that implements quantile-quantile plots. However, this is only useful, if we can transform the axes with `-log10()`. How to create custom made ggplot2 scales is explained [here](https://stat.ethz.ch/pipermail/r-help/2014-April/373728.html). 

```{r, fig.height = 5, fig.width = 5}

library(ggplot2)
theme_set(theme_bw())

## source: 
(gg <- ggplot(data = giant) + 
  geom_qq(aes(sample = P), distribution = stats::qunif)) +
    theme(aspect.ratio = 1)

# + scale_y_log10() + scale_x_log10()
# + coord_trans(x = "log10", y = "log10")
  
f_trans_mlog10 <- function(x) -log10(x)
f_trans_mlog10_inverse <- function(x) 10^(-x)

mlog10_trans <- scales::trans_new("-log10",
                                  transform = f_trans_mlog10,
                                  inverse = f_trans_mlog10_inverse)

gg + 
  scale_y_continuous(trans = mlog10_trans) + 
  scale_x_continuous(trans = mlog10_trans) + 
  geom_abline(intercept = 0, slope = 1) + 
  theme(aspect.ratio = 1)
```

### 3. ggplot2::geom_point

```{r, fig.height = 5, fig.width = 5}

N <- nrow(giant)
expected <- (1:N) / N - 1 / (2 * N)
giant <- giant %>% dplyr::arrange(P) %>% dplyr::mutate(P_expected = expected)

ggplot(data = giant) + 
  geom_point(aes(-log10(P_expected), -log10(P))) + 
  geom_abline(intercept = 0, slope = 1) + 
  theme(aspect.ratio = 1) + 
  xlab(expression(Expected ~ ~-log[10](italic(p)))) + ## from qqman::qq
  ylab(expression(Observed ~ ~-log[10](italic(p)))) ## from qqman::qq



## or with the scale_y_continous feature
ggplot(data = giant) +
geom_point(aes(P_expected, P)) + 
  scale_y_continuous(trans = mlog10_trans) + 
  scale_x_continuous(trans = mlog10_trans) + 
  geom_abline(intercept = 0, slope = 1) + 
  theme(aspect.ratio = 1) #+ ## square shaped
#  expand_limits(x = -log10(max(giant$P)), y = -log10(max(giant$P)))
  
```


### 4. ggGWAS::geom_gwas_qq

`ggplot(data = giant) + geom_gwas_qq(aes(y = P))`

Here is a whishlist of what a `geom_gwas_qq` function should be able to do:

- include correct labels (expected as x-label and observed as y-label)
- make sure color, group, facetting all works
- allow for the `raster` version (for faster plotting), hexbins (also for faster plotting) and Pvalue thresholding (removing the high Pvalue SNPs from the plot)
- if time: implement genomic inflation factor representation
- while we are at it: plotting box should be squared (`theme(aspect.ratio = 1)`)
 
<!-------
### 5. Ascii

https://stackoverflow.com/questions/14736556/ascii-plotting-functions-for-r/14742722#14742722

```{r, eval=FALSE, include=FALSE}
library(txtplot)

with(giant, txtplot(P_expected, P))
```
Kinda pointless, but worth a try. 

------->
 
## Ways of making a ggplot2-like Manhattan plot

### 1. qqman::manhattan


```{r, fig.height = 5, fig.width = 8}
qqman::manhattan(giant, chr = "CHR" , bp = "POS", p = "P")
```

### 2. ggplot2::geom_point

```{r, fig.height = 5, fig.width = 8}

## computing new x axis
giant <- giant %>% dplyr::arrange(CHR, POS) %>% dplyr::mutate(tmp = 1, cumsum.tmp = cumsum(tmp))

## calculating x axis location for chromosome label
med.dat <- giant %>% dplyr::group_by(CHR) %>% dplyr::summarise(median.x = median(cumsum.tmp))

ggplot(data = giant) + 
  geom_point(aes(x = cumsum.tmp, y = -log10(P), color = CHR %%2 == 0)) + ## create alternate coloring
  geom_hline(yintercept = -log10(5e-8)) + ## add horizontal line
  scale_x_continuous(breaks = med.dat$median.x, labels = med.dat$CHR) + ## add new x labels 
  guides(colour=FALSE) +  ## remove legend
  xlab("Chromosome") + 
  ylab(expression(-log[10](italic(p)))) + ## y label from qqman::qq
  scale_color_manual(values = c(gray(0.5), gray(0))) ## instead of colors, go for gray
  
```




### 3. ggGWAS::geom_gwas_manhattan

`ggplot(data = giant) + geom_gwas_manhattan(aes(pos = POS, chr = CHR, y = -log10(P)))`

Here is a whishlist of what a `geom_gwas_manhattan` function should be able to do:

- x axis spacing with space between chromosome and spaced as with position
- include correct y axis labels 
- make sure color, group, facetting all works
- allow for the `raster` version (for faster plotting) and Pvalue thresholding (removing the high Pvalue SNPs from the plot)
- implement coloring (two alternating colors)


## Useful links

### How to make a ggplot2 extension

- [Extending ggplot2 (vignette)](https://ggplot2.tidyverse.org/articles/extending-ggplot2.html#creating-a-new-geom)
- [Programming with ggplot2 (rubs)](https://rpubs.com/hadley/97970)
- [R help](https://www.rdocumentation.org/packages/ggplot2/versions/3.0.0/topics/ggplot2-ggproto)
- [here (wiki)](https://github.com/tidyverse/ggplot2/wiki/Creating-a-new-geom) (from 2010)
- [ggproto help](https://ggplot2.tidyverse.org/reference/ggplot2-ggproto.html)
- [Howto by R Peng](https://bookdown.org/rdpeng/RProgDA/building-new-graphical-elements.html)
- [Creating ggplot2 extensions (rud)](https://rud.is/books/creating-ggplot2-extensions/)
- [Creating stats and geoms (rud)](https://rud.is/b/2015/09/08/roll-your-own-stats-and-geoms-in-ggplot2-part-1-splines/)
- https://twitter.com/DrMowinckels/status/1115124147405901824

### Other ggplot2 extensions

- https://github.com/LCBC-UiO/ggseg
- https://www.garrickadenbuie.com/project/ggpomological/
- https://coolbutuseless.github.io/2019/03/19/geom_blurry-proof-of-concept/


## Misc

### Testing

How to test plots? 

One option is, to compare ggplot2 object data. In the example below, we are comparing two ggplot2 outputs, one created with `qplot` and one with `ggplot`. 

```
gg1 <- qplot(Sepal.Length, Petal.Length, data = iris)
gg2 <- ggplot(data = iris) + geom_point(aes(Sepal.Length, Petal.Length))
identical(gg1$data, gg2$data)
```
We can apply this to our package by creating the qqplot and manhattanplots manually by hand, and then comparing the to the function outputs.

Another option is to use https://github.com/lionel-/vdiffr



