<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>ggGWAS reasoning • ggGWAS</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha256-eZrrJcwDc/3uDhsdt61sL2oOBY362qM3lon1gyExkL0=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- sticky kit --><script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="ggGWAS reasoning">
<meta property="og:description" content="">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">ggGWAS</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.0.2.9000</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/gggwas-internals.html">ggGWAS Internals</a>
    </li>
    <li>
      <a href="../articles/gggwas-reasoning.html">ggGWAS reasoning</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>ggGWAS reasoning</h1>
                        <h4 class="author">Sina Rüeger</h4>
            
            <h4 class="date">2019-07-31</h4>
      
      
      <div class="hidden name"><code>gggwas-reasoning.Rmd</code></div>

    </div>

    
    
<p>Why bother with ggplot2 extension for GWAS summary statistic data visualisations?</p>
<p>This vignette lays out the <a href="#reasons">motivation</a> why having a ggplot2 extension is needed. But it also discusses existing <a href="#alternatives">alternatives</a> and ways to create a ggplot2-like <a href="#ways-of-making-a-ggplot2-like-qqplot-plot">Q-Q plot</a> and <a href="#ways-of-making-a-ggplot2-like-manhattan-plot">Manhattanplot</a>.</p>
<div id="gwas-data-visualisations" class="section level2">
<h2 class="hasAnchor">
<a href="#gwas-data-visualisations" class="anchor"></a>GWAS data visualisations</h2>
<p>There are two main data visualisations<a href="#fn1" class="footnote-ref" id="fnref1"><sup>1</sup></a> that are done after the GWAS is run<a href="#fn2" class="footnote-ref" id="fnref2"><sup>2</sup></a>:</p>
<ul>
<li><p><a href="https://en.wikipedia.org/wiki/Q%E2%80%93Q_plot">Q-Q plot</a>: checks the deviation of the P-value distribution from the uniform distribution.</p></li>
<li><p><a href="https://en.wikipedia.org/wiki/Manhattan_plot">Manhattan plot</a>: shows the P-value or summary statistic distribution along the chomosomal position.</p></li>
</ul>
<p><img src="https://user-images.githubusercontent.com/4454726/47022176-323fb600-d15d-11e8-892b-152816ce9574.png" alt="QQplot" height="249"></p>
<p>
<a href="https://commons.wikimedia.org/wiki/File:Manhattan_Plot.png#/media/File:Manhattan_Plot.png"><img src="https://upload.wikimedia.org/wikipedia/commons/1/12/Manhattan_Plot.png" alt="Manhattan Plot.png" width="640" height="249"></a><br>By M. Kamran Ikram et al - Ikram MK et al (2010) Four Novel Loci (19q13, 6q24, 12q24, and 5q14) Influence the Microcirculation In Vivo. PLoS Genet. 2010 Oct 28;6(10):e1001184. <a href="https://en.wikipedia.org/wiki/Digital_object_identifier" class="extiw" title="w:Digital object identifier">doi</a>:<a rel="nofollow" class="external text" href="https://doi.org/10.1371%2Fjournal.pgen.1001184.g001">10.1371/journal.pgen.1001184.g001</a>, <a href="https://creativecommons.org/licenses/by/2.5" title="Creative Commons Attribution 2.5">CC BY 2.5</a>, <a href="https://commons.wikimedia.org/w/index.php?curid=18056138">Link</a>
</p>
</div>
<div id="alternatives" class="section level2">
<h2 class="hasAnchor">
<a href="#alternatives" class="anchor"></a>Alternatives</h2>
<p>Both plotting types can be applied in R with various functions. Just to name a few, here are some links which have also served as inspiration:</p>
<ul>
<li>
<code>qqman</code>: R package for GWAS results: <a href="http://www.gettinggeneticsdone.com/2014/05/qqman-r-package-for-qq-and-manhattan-plots-for-gwas-results.html" class="uri">http://www.gettinggeneticsdone.com/2014/05/qqman-r-package-for-qq-and-manhattan-plots-for-gwas-results.html</a>
</li>
<li>
<code>ggman</code>: R package with ggplot2 wrapper with some neat hacks <a href="https://github.com/mkanai/ggman" class="uri">https://github.com/mkanai/ggman</a>
</li>
<li>Recipe to create Manhattanplot in R with <code>qqman</code>, <code>ggplot2</code> and <code>plotly</code>: <a href="https://www.r-graph-gallery.com/wp-content/uploads/2018/02/Manhattan_plot_in_R.html" class="uri">https://www.r-graph-gallery.com/wp-content/uploads/2018/02/Manhattan_plot_in_R.html</a>
</li>
<li>Recipe for QQplots in <code>lattice</code> (for historical reasons) <a href="https://genome.sph.umich.edu/wiki/Code_Sample:_Generating_QQ_Plots_in_R" class="uri">https://genome.sph.umich.edu/wiki/Code_Sample:_Generating_QQ_Plots_in_R</a>
</li>
<li>Fast QQplot implementation in <code>ramwas</code> package: <a href="https://rdrr.io/bioc/ramwas/man/qqplotFast.html" class="uri">https://rdrr.io/bioc/ramwas/man/qqplotFast.html</a>
</li>
<li>DYI <code>ggplot2</code> Manhattan plots: <a href="https://github.com/pcgoddard/Burchardlab_Tutorials/wiki/GGplot2-Manhattan-Plot-Function" class="uri">https://github.com/pcgoddard/Burchardlab_Tutorials/wiki/GGplot2-Manhattan-Plot-Function</a>
</li>
<li>
<code>ggbio</code> package, a toolkit for genomic data (<code>plotGrandLinear</code> is an implementation for a Manhattan plot, uses <code>GRanges</code> objects, ggplot2 wrapper): <a href="https://bioconductor.org/packages/release/bioc/html/ggbio.html" class="uri">https://bioconductor.org/packages/release/bioc/html/ggbio.html</a>
</li>
<li>
<code>ggman</code>, wrapper (but no ggplot2 extension) for Manhattan plots: <a href="https://github.com/drveera/ggman" class="uri">https://github.com/drveera/ggman</a>
</li>
<li><a href="https://www.gettinggeneticsdone.com/2010/07/qq-plots-of-p-values-in-r-using-base.html" class="uri">https://www.gettinggeneticsdone.com/2010/07/qq-plots-of-p-values-in-r-using-base.html</a></li>
</ul>
<p>An often used package is <a href="https://rdrr.io/cran/qqman/man/qq.html">qqman</a>, that has a function for Q-Q plots <code><a href="https://www.rdocumentation.org/packages/qqman/topics/qq">qqman::qq</a></code> and Manhattan plots <code><a href="https://www.rdocumentation.org/packages/qqman/topics/manhattan">qqman::manhattan</a></code>.</p>
<p>And there are various ggplot2 wrappers (<code>ggbio</code>, <code>ggman</code>), but not actual extensions.</p>
</div>
<div id="reasons" class="section level2">
<h2 class="hasAnchor">
<a href="#reasons" class="anchor"></a>Reasons</h2>
<p>The problem is two fold:</p>
<ol style="list-style-type: decimal">
<li>GWAS summary statistics usually contains around 1 mio of rows, so plotting becomes a lengthy process.</li>
<li>All existing packages use the base plotting functions, making it difficult to apply any functionalities that <code>ggplot2</code> can (e.g. facetting, colouring).</li>
</ol>
<p>The idea is therefore to create a ggplot2 extension that would facilitate all this: 1) fast plotting (through hexagons and filtering) and 2) building <code>ggplot2</code> geoms and stats.</p>
<p>By writing a ggplot2 extension, we can inherit lots of the default ggplot2 functionalities and shorten the input.</p>
</div>
<div id="ways-of-making-a-ggplot2-like-q-q-plot-plot" class="section level2">
<h2 class="hasAnchor">
<a href="#ways-of-making-a-ggplot2-like-q-q-plot-plot" class="anchor"></a>Ways of making a ggplot2-like Q-Q plot plot</h2>
<p>Let’s say we have GWAS summary statistics for a number of SNPs. We call this data <code>gwas.summarystats</code>: for a number of SNPs listed by row, we know the SNP identifier (<code>SNP</code>) and the P-value (<code>P</code>).</p>
<pre><code>SNP     P
rs3342  1e-2
rs83    1e-2
...     ...</code></pre>
<p>For illustration purposes, we will use the summary statistics dataset <code><a href="https://www.rdocumentation.org/packages/GWAS.utils/topics/giant">GWAS.utils::giant</a></code>.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb2-1" title="1"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(GWAS.utils)</a>
<a class="sourceLine" id="cb2-2" title="2"><span class="co">#skimr::skim(giant)</span></a>
<a class="sourceLine" id="cb2-3" title="3"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/summary">summary</a></span>(giant)</a>
<a class="sourceLine" id="cb2-4" title="4"><span class="co">#&gt;       CHR             POS                SNP            Tested_Allele     </span></a>
<a class="sourceLine" id="cb2-5" title="5"><span class="co">#&gt;  Min.   : 1.00   Min.   :   220149   Length:10000       Length:10000      </span></a>
<a class="sourceLine" id="cb2-6" title="6"><span class="co">#&gt;  1st Qu.: 4.00   1st Qu.: 33810156   Class :character   Class :character  </span></a>
<a class="sourceLine" id="cb2-7" title="7"><span class="co">#&gt;  Median : 8.00   Median : 71004162   Mode  :character   Mode  :character  </span></a>
<a class="sourceLine" id="cb2-8" title="8"><span class="co">#&gt;  Mean   : 8.55   Mean   : 79743155                                        </span></a>
<a class="sourceLine" id="cb2-9" title="9"><span class="co">#&gt;  3rd Qu.:12.00   3rd Qu.:114753146                                        </span></a>
<a class="sourceLine" id="cb2-10" title="10"><span class="co">#&gt;  Max.   :22.00   Max.   :248476804                                        </span></a>
<a class="sourceLine" id="cb2-11" title="11"><span class="co">#&gt;  Other_Allele       Freq_Tested_Allele_in_HRS      BETA           </span></a>
<a class="sourceLine" id="cb2-12" title="12"><span class="co">#&gt;  Length:10000       Min.   :0.01023           Min.   :-4.980e-02  </span></a>
<a class="sourceLine" id="cb2-13" title="13"><span class="co">#&gt;  Class :character   1st Qu.:0.19750           1st Qu.:-2.800e-03  </span></a>
<a class="sourceLine" id="cb2-14" title="14"><span class="co">#&gt;  Mode  :character   Median :0.45710           Median :-1.000e-04  </span></a>
<a class="sourceLine" id="cb2-15" title="15"><span class="co">#&gt;                     Mean   :0.48108           Mean   :-4.439e-05  </span></a>
<a class="sourceLine" id="cb2-16" title="16"><span class="co">#&gt;                     3rd Qu.:0.75913           3rd Qu.: 2.600e-03  </span></a>
<a class="sourceLine" id="cb2-17" title="17"><span class="co">#&gt;                     Max.   :0.98977           Max.   : 4.480e-02  </span></a>
<a class="sourceLine" id="cb2-18" title="18"><span class="co">#&gt;        SE                 P                N         </span></a>
<a class="sourceLine" id="cb2-19" title="19"><span class="co">#&gt;  Min.   :0.001600   Min.   :0.0000   Min.   :498178  </span></a>
<a class="sourceLine" id="cb2-20" title="20"><span class="co">#&gt;  1st Qu.:0.001800   1st Qu.:0.0460   1st Qu.:680198  </span></a>
<a class="sourceLine" id="cb2-21" title="21"><span class="co">#&gt;  Median :0.002100   Median :0.2600   Median :688732  </span></a>
<a class="sourceLine" id="cb2-22" title="22"><span class="co">#&gt;  Mean   :0.002596   Mean   :0.3406   Mean   :683795  </span></a>
<a class="sourceLine" id="cb2-23" title="23"><span class="co">#&gt;  3rd Qu.:0.002800   3rd Qu.:0.6000   3rd Qu.:691571  </span></a>
<a class="sourceLine" id="cb2-24" title="24"><span class="co">#&gt;  Max.   :0.010600   Max.   :1.0000   Max.   :795594</span></a></code></pre></div>
<div id="qqmanqq" class="section level3">
<h3 class="hasAnchor">
<a href="#qqmanqq" class="anchor"></a>1. qqman::qq</h3>
<p><code><a href="https://www.rdocumentation.org/packages/qqman/topics/qq">qqman::qq</a></code> is straightforward to use: just pass on the P-value column.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb3-1" title="1">qqman<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/qqman/topics/qq">qq</a></span>(giant<span class="op">$</span>P)</a></code></pre></div>
<p><img src="gggwas-reasoning_files/figure-html/unnamed-chunk-2-1.png" width="480"></p>
</div>
<div id="ggplot2geom_qq" class="section level3">
<h3 class="hasAnchor">
<a href="#ggplot2geom_qq" class="anchor"></a>2. ggplot2::geom_qq</h3>
<p>There is a <a href="https://ggplot2.tidyverse.org/reference/geom_qq.html"><code>geom_qq</code></a> in ggplot2 that implements quantile-quantile plots. However, this is only useful, if we can transform the axes with <code>-log10()</code>. How to create custom made ggplot2 scales is explained <a href="https://stat.ethz.ch/pipermail/r-help/2014-April/373728.html">here</a>.</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb4-1" title="1"></a>
<a class="sourceLine" id="cb4-2" title="2"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(ggplot2)</a>
<a class="sourceLine" id="cb4-3" title="3"><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/theme_get.html">theme_set</a></span>(<span class="kw"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_bw</a></span>())</a>
<a class="sourceLine" id="cb4-4" title="4"></a>
<a class="sourceLine" id="cb4-5" title="5"><span class="co">## source: </span></a>
<a class="sourceLine" id="cb4-6" title="6">(gg &lt;-<span class="st"> </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span>(<span class="dt">data =</span> giant) <span class="op">+</span><span class="st"> </span></a>
<a class="sourceLine" id="cb4-7" title="7"><span class="st">  </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/geom_qq.html">geom_qq</a></span>(<span class="kw"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(<span class="dt">sample =</span> P), <span class="dt">distribution =</span> stats<span class="op">::</span>qunif)) <span class="op">+</span></a>
<a class="sourceLine" id="cb4-8" title="8"><span class="st">    </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span>(<span class="dt">aspect.ratio =</span> <span class="dv">1</span>)</a></code></pre></div>
<p><img src="gggwas-reasoning_files/figure-html/unnamed-chunk-3-1.png" width="480"></p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb5-1" title="1"></a>
<a class="sourceLine" id="cb5-2" title="2"><span class="co"># + scale_y_log10() + scale_x_log10()</span></a>
<a class="sourceLine" id="cb5-3" title="3"><span class="co"># + coord_trans(x = "log10", y = "log10")</span></a>
<a class="sourceLine" id="cb5-4" title="4">  </a>
<a class="sourceLine" id="cb5-5" title="5">f_trans_mlog10 &lt;-<span class="st"> </span><span class="cf">function</span>(x) <span class="op">-</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/Log">log10</a></span>(x)</a>
<a class="sourceLine" id="cb5-6" title="6">f_trans_mlog10_inverse &lt;-<span class="st"> </span><span class="cf">function</span>(x) <span class="dv">10</span><span class="op">^</span>(<span class="op">-</span>x)</a>
<a class="sourceLine" id="cb5-7" title="7"></a>
<a class="sourceLine" id="cb5-8" title="8">mlog10_trans &lt;-<span class="st"> </span>scales<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/scales/topics/trans_new">trans_new</a></span>(<span class="st">"-log10"</span>,</a>
<a class="sourceLine" id="cb5-9" title="9">                                  <span class="dt">transform =</span> f_trans_mlog10,</a>
<a class="sourceLine" id="cb5-10" title="10">                                  <span class="dt">inverse =</span> f_trans_mlog10_inverse)</a>
<a class="sourceLine" id="cb5-11" title="11"></a>
<a class="sourceLine" id="cb5-12" title="12">gg <span class="op">+</span><span class="st"> </span></a>
<a class="sourceLine" id="cb5-13" title="13"><span class="st">  </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html">scale_y_continuous</a></span>(<span class="dt">trans =</span> mlog10_trans) <span class="op">+</span><span class="st"> </span></a>
<a class="sourceLine" id="cb5-14" title="14"><span class="st">  </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html">scale_x_continuous</a></span>(<span class="dt">trans =</span> mlog10_trans) <span class="op">+</span><span class="st"> </span></a>
<a class="sourceLine" id="cb5-15" title="15"><span class="st">  </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html">geom_abline</a></span>(<span class="dt">intercept =</span> <span class="dv">0</span>, <span class="dt">slope =</span> <span class="dv">1</span>) <span class="op">+</span><span class="st"> </span></a>
<a class="sourceLine" id="cb5-16" title="16"><span class="st">  </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span>(<span class="dt">aspect.ratio =</span> <span class="dv">1</span>)</a></code></pre></div>
<p><img src="gggwas-reasoning_files/figure-html/unnamed-chunk-3-2.png" width="480"></p>
</div>
<div id="ggplot2geom_point" class="section level3">
<h3 class="hasAnchor">
<a href="#ggplot2geom_point" class="anchor"></a>3. ggplot2::geom_point</h3>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb6-1" title="1"></a>
<a class="sourceLine" id="cb6-2" title="2">N &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/nrow">nrow</a></span>(giant)</a>
<a class="sourceLine" id="cb6-3" title="3">expected &lt;-<span class="st"> </span>(<span class="dv">1</span><span class="op">:</span>N) <span class="op">/</span><span class="st"> </span>N <span class="op">-</span><span class="st"> </span><span class="dv">1</span> <span class="op">/</span><span class="st"> </span>(<span class="dv">2</span> <span class="op">*</span><span class="st"> </span>N)</a>
<a class="sourceLine" id="cb6-4" title="4">giant &lt;-<span class="st"> </span>giant <span class="op">%&gt;%</span><span class="st"> </span>dplyr<span class="op">::</span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/arrange.html">arrange</a></span>(P) <span class="op">%&gt;%</span><span class="st"> </span>dplyr<span class="op">::</span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span>(<span class="dt">P_expected =</span> expected)</a>
<a class="sourceLine" id="cb6-5" title="5"></a>
<a class="sourceLine" id="cb6-6" title="6"><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span>(<span class="dt">data =</span> giant) <span class="op">+</span><span class="st"> </span></a>
<a class="sourceLine" id="cb6-7" title="7"><span class="st">  </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span>(<span class="kw"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(<span class="op">-</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/Log">log10</a></span>(P_expected), <span class="op">-</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/Log">log10</a></span>(P))) <span class="op">+</span><span class="st"> </span></a>
<a class="sourceLine" id="cb6-8" title="8"><span class="st">  </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html">geom_abline</a></span>(<span class="dt">intercept =</span> <span class="dv">0</span>, <span class="dt">slope =</span> <span class="dv">1</span>) <span class="op">+</span><span class="st"> </span></a>
<a class="sourceLine" id="cb6-9" title="9"><span class="st">  </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span>(<span class="dt">aspect.ratio =</span> <span class="dv">1</span>) <span class="op">+</span><span class="st"> </span></a>
<a class="sourceLine" id="cb6-10" title="10"><span class="st">  </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/labs.html">xlab</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/expression">expression</a></span>(Expected <span class="op">~</span><span class="st"> </span><span class="er">~</span><span class="op">-</span>log[<span class="dv">10</span>](<span class="kw"><a href="https://www.rdocumentation.org/packages/grDevices/topics/plotmath">italic</a></span>(p)))) <span class="op">+</span><span class="st"> </span><span class="co">## from qqman::qq</span></a>
<a class="sourceLine" id="cb6-11" title="11"><span class="st">  </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ylab</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/expression">expression</a></span>(Observed <span class="op">~</span><span class="st"> </span><span class="er">~</span><span class="op">-</span>log[<span class="dv">10</span>](<span class="kw"><a href="https://www.rdocumentation.org/packages/grDevices/topics/plotmath">italic</a></span>(p)))) <span class="co">## from qqman::qq</span></a></code></pre></div>
<p><img src="gggwas-reasoning_files/figure-html/unnamed-chunk-4-1.png" width="480"></p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb7-1" title="1"></a>
<a class="sourceLine" id="cb7-2" title="2"></a>
<a class="sourceLine" id="cb7-3" title="3"></a>
<a class="sourceLine" id="cb7-4" title="4"><span class="co">## or with the scale_y_continous feature</span></a>
<a class="sourceLine" id="cb7-5" title="5"><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span>(<span class="dt">data =</span> giant) <span class="op">+</span></a>
<a class="sourceLine" id="cb7-6" title="6"><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span>(<span class="kw"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(P_expected, P)) <span class="op">+</span><span class="st"> </span></a>
<a class="sourceLine" id="cb7-7" title="7"><span class="st">  </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html">scale_y_continuous</a></span>(<span class="dt">trans =</span> mlog10_trans) <span class="op">+</span><span class="st"> </span></a>
<a class="sourceLine" id="cb7-8" title="8"><span class="st">  </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html">scale_x_continuous</a></span>(<span class="dt">trans =</span> mlog10_trans) <span class="op">+</span><span class="st"> </span></a>
<a class="sourceLine" id="cb7-9" title="9"><span class="st">  </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html">geom_abline</a></span>(<span class="dt">intercept =</span> <span class="dv">0</span>, <span class="dt">slope =</span> <span class="dv">1</span>) <span class="op">+</span><span class="st"> </span></a>
<a class="sourceLine" id="cb7-10" title="10"><span class="st">  </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span>(<span class="dt">aspect.ratio =</span> <span class="dv">1</span>) <span class="co">#+ ## square shaped</span></a></code></pre></div>
<p><img src="gggwas-reasoning_files/figure-html/unnamed-chunk-4-2.png" width="480"></p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb8-1" title="1"><span class="co">#  expand_limits(x = -log10(max(giant$P)), y = -log10(max(giant$P)))</span></a>
<a class="sourceLine" id="cb8-2" title="2">  </a></code></pre></div>
</div>
<div id="gggwasgeom_gwas_qq" class="section level3">
<h3 class="hasAnchor">
<a href="#gggwasgeom_gwas_qq" class="anchor"></a>4. ggGWAS::geom_gwas_qq</h3>
<p><code>ggplot(data = giant) + geom_gwas_qq(aes(y = P))</code></p>
<p>Here is a whishlist of what a <code>geom_gwas_qq</code> function should be able to do:</p>
<ul>
<li>include correct labels (expected as x-label and observed as y-label)</li>
<li>make sure color, group, facetting all works</li>
<li>allow for the <code>raster</code> version (for faster plotting), hexbins (also for faster plotting) and Pvalue thresholding (removing the high Pvalue SNPs from the plot)</li>
<li>if time: implement genomic inflation factor representation</li>
<li>while we are at it: plotting box should be squared (<code><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme(aspect.ratio = 1)</a></code>)</li>
</ul>
<!-------
### 5. Ascii

https://stackoverflow.com/questions/14736556/ascii-plotting-functions-for-r/14742722#14742722


Kinda pointless, but worth a try. 

------->
</div>
</div>
<div id="ways-of-making-a-ggplot2-like-manhattan-plot" class="section level2">
<h2 class="hasAnchor">
<a href="#ways-of-making-a-ggplot2-like-manhattan-plot" class="anchor"></a>Ways of making a ggplot2-like Manhattan plot</h2>
<div id="qqmanmanhattan" class="section level3">
<h3 class="hasAnchor">
<a href="#qqmanmanhattan" class="anchor"></a>1. qqman::manhattan</h3>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb9-1" title="1">qqman<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/qqman/topics/manhattan">manhattan</a></span>(giant, <span class="dt">chr =</span> <span class="st">"CHR"</span> , <span class="dt">bp =</span> <span class="st">"POS"</span>, <span class="dt">p =</span> <span class="st">"P"</span>)</a></code></pre></div>
<p><img src="gggwas-reasoning_files/figure-html/unnamed-chunk-6-1.png" width="768"></p>
</div>
<div id="ggplot2geom_point-1" class="section level3">
<h3 class="hasAnchor">
<a href="#ggplot2geom_point-1" class="anchor"></a>2. ggplot2::geom_point</h3>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb10-1" title="1"></a>
<a class="sourceLine" id="cb10-2" title="2"><span class="co">## computing new x axis</span></a>
<a class="sourceLine" id="cb10-3" title="3">giant &lt;-<span class="st"> </span>giant <span class="op">%&gt;%</span><span class="st"> </span>dplyr<span class="op">::</span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/arrange.html">arrange</a></span>(CHR, POS) <span class="op">%&gt;%</span><span class="st"> </span>dplyr<span class="op">::</span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span>(<span class="dt">tmp =</span> <span class="dv">1</span>, <span class="dt">cumsum.tmp =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/cumsum">cumsum</a></span>(tmp))</a>
<a class="sourceLine" id="cb10-4" title="4"></a>
<a class="sourceLine" id="cb10-5" title="5"><span class="co">## calculating x axis location for chromosome label</span></a>
<a class="sourceLine" id="cb10-6" title="6">med.dat &lt;-<span class="st"> </span>giant <span class="op">%&gt;%</span><span class="st"> </span>dplyr<span class="op">::</span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/group_by.html">group_by</a></span>(CHR) <span class="op">%&gt;%</span><span class="st"> </span>dplyr<span class="op">::</span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/summarise.html">summarise</a></span>(<span class="dt">median.x =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/stats/topics/median">median</a></span>(cumsum.tmp))</a>
<a class="sourceLine" id="cb10-7" title="7"></a>
<a class="sourceLine" id="cb10-8" title="8"><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span>(<span class="dt">data =</span> giant) <span class="op">+</span><span class="st"> </span></a>
<a class="sourceLine" id="cb10-9" title="9"><span class="st">  </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span>(<span class="kw"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(<span class="dt">x =</span> cumsum.tmp, <span class="dt">y =</span> <span class="op">-</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/Log">log10</a></span>(P), <span class="dt">color =</span> CHR <span class="op">%%</span><span class="dv">2</span> <span class="op">==</span><span class="st"> </span><span class="dv">0</span>)) <span class="op">+</span><span class="st"> </span><span class="co">## create alternate coloring</span></a>
<a class="sourceLine" id="cb10-10" title="10"><span class="st">  </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html">geom_hline</a></span>(<span class="dt">yintercept =</span> <span class="op">-</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/Log">log10</a></span>(<span class="fl">5e-8</span>)) <span class="op">+</span><span class="st"> </span><span class="co">## add horizontal line</span></a>
<a class="sourceLine" id="cb10-11" title="11"><span class="st">  </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html">scale_x_continuous</a></span>(<span class="dt">breaks =</span> med.dat<span class="op">$</span>median.x, <span class="dt">labels =</span> med.dat<span class="op">$</span>CHR) <span class="op">+</span><span class="st"> </span><span class="co">## add new x labels </span></a>
<a class="sourceLine" id="cb10-12" title="12"><span class="st">  </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/guides.html">guides</a></span>(<span class="dt">colour=</span><span class="ot">FALSE</span>) <span class="op">+</span><span class="st">  </span><span class="co">## remove legend</span></a>
<a class="sourceLine" id="cb10-13" title="13"><span class="st">  </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/labs.html">xlab</a></span>(<span class="st">"Chromosome"</span>) <span class="op">+</span><span class="st"> </span></a>
<a class="sourceLine" id="cb10-14" title="14"><span class="st">  </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ylab</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/expression">expression</a></span>(<span class="op">-</span>log[<span class="dv">10</span>](<span class="kw"><a href="https://www.rdocumentation.org/packages/grDevices/topics/plotmath">italic</a></span>(p)))) <span class="op">+</span><span class="st"> </span><span class="co">## y label from qqman::qq</span></a>
<a class="sourceLine" id="cb10-15" title="15"><span class="st">  </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html">scale_color_manual</a></span>(<span class="dt">values =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/grDevices/topics/gray">gray</a></span>(<span class="fl">0.5</span>), <span class="kw"><a href="https://www.rdocumentation.org/packages/grDevices/topics/gray">gray</a></span>(<span class="dv">0</span>))) <span class="co">## instead of colors, go for gray</span></a></code></pre></div>
<p><img src="gggwas-reasoning_files/figure-html/unnamed-chunk-7-1.png" width="768"></p>
</div>
<div id="gggwasgeom_gwas_manhattan" class="section level3">
<h3 class="hasAnchor">
<a href="#gggwasgeom_gwas_manhattan" class="anchor"></a>3. ggGWAS::geom_gwas_manhattan</h3>
<p><code>ggplot(data = giant) + geom_gwas_manhattan(aes(pos = POS, chr = CHR, y = -log10(P)))</code></p>
<p>Here is a whishlist of what a <code>geom_gwas_manhattan</code> function should be able to do:</p>
<ul>
<li>x axis spacing with space between chromosome and spaced as with position</li>
<li>include correct y axis labels</li>
<li>make sure color, group, facetting all works</li>
<li>allow for the <code>raster</code> version (for faster plotting) and Pvalue thresholding (removing the high Pvalue SNPs from the plot)</li>
<li>implement coloring (two alternating colors)</li>
</ul>
</div>
</div>
<div id="useful-links" class="section level2">
<h2 class="hasAnchor">
<a href="#useful-links" class="anchor"></a>Useful links</h2>
<div id="how-to-make-a-ggplot2-extension" class="section level3">
<h3 class="hasAnchor">
<a href="#how-to-make-a-ggplot2-extension" class="anchor"></a>How to make a ggplot2 extension</h3>
<ul>
<li><a href="https://ggplot2.tidyverse.org/articles/extending-ggplot2.html#creating-a-new-geom">Extending ggplot2 (vignette)</a></li>
<li><a href="https://ggplot2.tidyverse.org/reference/ggplot2-ggproto.html">Base ggproto classes for ggplot2 (help)</a></li>
<li><a href="https://rpubs.com/hadley/97970">Programming with ggplot2 (rubs)</a></li>
<li><a href="https://www.rdocumentation.org/packages/ggplot2/versions/3.0.0/topics/ggplot2-ggproto">R help</a></li>
<li>
<a href="https://github.com/tidyverse/ggplot2/wiki/Creating-a-new-geom">here (wiki)</a> (from 2010)</li>
<li><a href="https://ggplot2.tidyverse.org/reference/ggplot2-ggproto.html">ggproto help</a></li>
<li><a href="https://bookdown.org/rdpeng/RProgDA/building-new-graphical-elements.html">Howto by R Peng</a></li>
<li><a href="https://rud.is/books/creating-ggplot2-extensions/">Creating ggplot2 extensions (rud)</a></li>
<li><a href="https://rud.is/b/2015/09/08/roll-your-own-stats-and-geoms-in-ggplot2-part-1-splines/">Creating stats and geoms (rud)</a></li>
<li><a href="https://twitter.com/DrMowinckels/status/1115124147405901824" class="uri">https://twitter.com/DrMowinckels/status/1115124147405901824</a></li>
</ul>
</div>
<div id="other-ggplot2-extensions" class="section level3">
<h3 class="hasAnchor">
<a href="#other-ggplot2-extensions" class="anchor"></a>Other ggplot2 extensions</h3>
<ul>
<li><a href="https://github.com/LCBC-UiO/ggseg" class="uri">https://github.com/LCBC-UiO/ggseg</a></li>
<li><a href="https://www.garrickadenbuie.com/project/ggpomological/" class="uri">https://www.garrickadenbuie.com/project/ggpomological/</a></li>
<li><a href="https://coolbutuseless.github.io/2019/03/19/geom_blurry-proof-of-concept/" class="uri">https://coolbutuseless.github.io/2019/03/19/geom_blurry-proof-of-concept/</a></li>
</ul>
</div>
</div>
<div id="misc" class="section level2">
<h2 class="hasAnchor">
<a href="#misc" class="anchor"></a>Misc</h2>
<div id="testing" class="section level3">
<h3 class="hasAnchor">
<a href="#testing" class="anchor"></a>Testing</h3>
<p>How to test plots?</p>
<p>One option is, to compare ggplot2 object data. In the example below, we are comparing two ggplot2 outputs, one created with <code>qplot</code> and one with <code>ggplot</code>.</p>
<pre><code>gg1 &lt;- qplot(Sepal.Length, Petal.Length, data = iris)
gg2 &lt;- ggplot(data = iris) + geom_point(aes(Sepal.Length, Petal.Length))
identical(gg1$data, gg2$data)</code></pre>
<p>We can apply this to our package by creating the qqplot and manhattanplots manually by hand, and then comparing the to the function outputs.</p>
<p>Another option is to use <a href="https://github.com/lionel-/vdiffr" class="uri">https://github.com/lionel-/vdiffr</a></p>
</div>
</div>
<div class="footnotes">
<hr>
<ol>
<li id="fn1"><p>There are also other tools used, such as locuszoom plots, but these are harder to implement because they need access to external information, and we will skip these for now.<a href="#fnref1" class="footnote-back">↩</a></p></li>
<li id="fn2"><p>We will not discuss their usefulness here.<a href="#fnref2" class="footnote-back">↩</a></p></li>
</ol>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#gwas-data-visualisations">GWAS data visualisations</a></li>
      <li><a href="#alternatives">Alternatives</a></li>
      <li><a href="#reasons">Reasons</a></li>
      <li><a href="#ways-of-making-a-ggplot2-like-q-q-plot-plot">Ways of making a ggplot2-like Q-Q plot plot</a></li>
      <li><a href="#ways-of-making-a-ggplot2-like-manhattan-plot">Ways of making a ggplot2-like Manhattan plot</a></li>
      <li><a href="#useful-links">Useful links</a></li>
      <li><a href="#misc">Misc</a></li>
      </ul>
</div>
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by Sina Rüeger.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.3.0.</p>
</div>
      </footer>
</div>

  

  </body>
</html>
